attempt to fix #5264

jangko opened the pull request.

Araq commented on compiler/semtypes.nim:
```diff
@@ -1042,6 +1043,35 @@ proc semBlockType(c: PContext, n: PNode, prev: PType): PType =
 proc semGenericParamInInvocation(c: PContext, n: PNode): PType =
   result = semTypeNode(c, n, nil)
 
+proc semRecordTypeAux(c: PContext, n: PNode, check: var IntSet) =
```
That looks like you copied it from somewhere. ;-) You can make the original a template perhaps and instantiate it here. Also your code is wrong ``case`` in objects can introduce different AST kinds (which is why the internal structure is so weird to begin with...).

jangko commented:
I have modified semRecordNodeAux a little bit, removed the unnecessary semRecordTypeAux, and add more test covering object with case.
Is there any other things I missed?
Actually I discover another bug during investigation: #5521

Araq commented on compiler/semtypes.nim:
```diff
@@ -1104,6 +1129,14 @@ proc semGeneric(c: PContext, n: PNode, s: PSym, prev: PType): PType =
+  # special check for generic object with
+  # parameterized parent
+  if result.kind == tyGenericInst:
+    var t = result.lastSon()
```
use something like this here:

```nim

let t = result.skipTypes(abstractPtrs)
if t != result and t.kind == tyObject and t.sons[0] != nil:
  semObjectType(c, n, t)
```

Araq commented on compiler/semtypes.nim:
```diff
@@ -1042,6 +1050,23 @@ proc semBlockType(c: PContext, n: PNode, prev: PType): PType =
 proc semGenericParamInInvocation(c: PContext, n: PNode): PType =
   result = semTypeNode(c, n, nil)
 
+proc semObjectType(c: PContext, n: PNode, t: PType) =
```
name this ``semObjectTypeForInheritedGenericInst`` since it's only used for this case.

Araq commented:
Looks good now, thanks! Will merge once AppVeyor has completed.

Araq commented on compiler/semtypes.nim:
```diff
@@ -1042,6 +1050,23 @@ proc semBlockType(c: PContext, n: PNode, prev: PType): PType =
+  var check = initIntSet()
+  var realBase = t.sons[0]
+  var pos = 0
+  var base = skipTypesOrNil(realBase, skipPtrs)
```
I think ``realBase`` and ``base`` can be ``let`` here. No need to change it, just a friendly note for the next time: It's not necessary to copy the compiler's code style when it is a leftover from the Pascal days. :-)

jangko replied:
😄 👍
